src/components/fields/ArrayField.js

```diff
@@ -88,68 +94,157 @@ class ArrayField extends Component {
   };
 
   render() {
+    const {schema} = this.props;
+    if (isFixedItems(schema)) {
+      return this.renderFixedArray();
+    }
+    if (isMultiSelect(schema)) {
+      return this.renderMultiSelect();
+    }
+    return this.renderNormalArray();
+  }
+
+  renderNormalArray() {
     const {schema, uiSchema, errorSchema, idSchema, name} = this.props;
     const title = schema.title || name;
     const {items} = this.state;
-    const {fields, definitions} = this.props.registry;
-    const {SchemaField} = fields;
+    const {definitions} = this.props.registry;
     const itemsSchema = retrieveSchema(schema.items, definitions);
-    if (isMultiSelect(schema)) {
-      return (
-        <SelectWidget
+
+    return (
+      <fieldset
+          className={`field field-array field-array-of-${itemsSchema.type}`}>
+        {title ? <legend>{title}</legend> : null}
+        {schema.description ?
+            <div className="field-description">{schema.description}</div> : null}
+        <div className="row array-item-list">{
+          items.map((item, index) => {
+            const itemErrorSchema = errorSchema ? errorSchema[index] : undefined;
+            const itemIdPrefix = idSchema.id + "_" + index;
+            const itemIdSchema = toIdSchema(itemsSchema, itemIdPrefix, definitions);
+            return this.renderArrayFieldItem({
+              index, itemSchema: itemsSchema, itemIdSchema, itemErrorSchema,
+              itemData: items[index],
+              itemUiSchema: uiSchema.items
+            });
+          })
+        }</div>
+        <AddButton onClick={this.onAddClick}/>
+      </fieldset>
+    );
+  }
+
+  renderMultiSelect() {
+    const {schema, idSchema, name} = this.props;
+    const title = schema.title || name;
+    const {items} = this.state;
+    const {definitions} = this.props.registry;
+    const itemsSchema = retrieveSchema(schema.items, definitions);
+    return (
+      <SelectWidget
           id={idSchema && idSchema.id}
           multiple
           onChange={this.onSelectChange}
           options={optionsList(itemsSchema)}
           schema={schema}
           title={title}
           value={items}
-        />
-      );
-    }
+      />
+    );
+  }
 
+  renderFixedArray() {
+    const {schema, uiSchema, errorSchema, idSchema, name} = this.props;
+    const title = schema.title || name;
+    const {items} = this.state;
+    const {definitions} = this.props.registry;
+    const itemSchemas = schema.items.map(item => retrieveSchema(item, definitions));
+    const additionalSchema = allowAdditionalItems(schema) ? retrieveSchema(schema.additionalItems, definitions) : null;
+    const additionalItems = additionalSchema && items && (items.length > itemSchemas.length) && items.slice(itemSchemas.length);
```

Comment:
Nit: Please try to keep lines <= 80 columns. Create intermediary consts if that helps.